packages/react/src/ReactFetch.js

```diff
@@ -0,0 +1,130 @@
+  // IF YOU CHANGE THIS UPDATE THE simpleCacheKey ABOVE.
+  return JSON.stringify([
+    request.method,
+    Array.from(request.headers.entries()),
+    request.mode,
+    request.redirect,
+    request.credentials,
+    request.referrer,
+    request.referrerPolicy,
+    request.integrity,
+  ]);
+}
+
+if (enableCache && enableFetchInstrumentation) {
+  if (typeof fetch === 'function') {
+    const originalFetch = fetch;
+    try {
+      // eslint-disable-next-line no-native-reassign
+      fetch = function fetch(
+        resource: URL | RequestInfo,
+        options?: RequestOptions,
+      ) {
+        const dispatcher = ReactCurrentCache.current;
+        if (!dispatcher) {
+          // We're outside a cached scope.
+          return originalFetch(resource, options);
+        }
+        if (
+          options &&
+          options.signal &&
+          options.signal !== dispatcher.getCacheSignal()
+        ) {
+          // If we're passed a signal that is not ours, then we assume that
+          // someone else controls the lifetime of this object and opts out of
+          // caching. It's effectively the opt-out mechanism.
+          // Ideally we should be able to check this on the Request but
+          // it always gets initialized with its own signal so we don't
+          // know if it's supposed to override - unless we also override the
+          // Request constructor.
+          return originalFetch(resource, options);
+        }
+        // Normalize the Request
+        let url: string;
+        let cacheKey: string;
+        if (typeof resource === 'string' && !options) {
+          // Fast path.
+          cacheKey = simpleCacheKey;
+          url = resource;
+        } else {
+          // Normalize the request.
+          const request = new Request(resource, options);
+          if (
+            (request.method !== 'GET' && request.method !== 'HEAD') ||
+            // $FlowFixMe: keepalive is real
+            request.keepalive
+          ) {
+            // We currently don't dedupe requests that might have side-effects. Those
+            // have to be explicitly cached. We assume that the request doesn't have a
+            // body if it's GET or HEAD.
+            // keepalive gets treated the same as if you passed a custom cache signal.
+            return originalFetch(resource, options);
+          }
+          cacheKey = generateCacheKey(request);
+          url = request.url;
+        }
+        const cache = dispatcher.getCacheForType(createFetchCache);
+        const cacheEntries = cache.get(url);
+        let match;
+        if (cacheEntries === undefined) {
+          // We pass the original arguments here in case normalizing the Request
+          // doesn't include all the options in this environment.
+          match = originalFetch(resource, options);
+          cache.set(url, [cacheKey, match]);
+        } else {
+          // We use an array as the inner data structure since it's lighter and
```

Comment:
lighter than both {} and Map? I know those use arrays internally for small sizes right?

Reply:
Didn't benchmark yet but maybe yea.